File: apps/web/src/types/comparable.ts
export type ComparableType = "sale" | "rental";
export type PropertyComparable = {
  comparable_id: string; subject_property_id: string; comparable_type: ComparableType;
  source_name: string; source_url: string; source_record_id?: string; address: string;
  latitude?: number; longitude?: number; distance_miles?: number; property_type: string; property_subtype?: string;
  unit_count?: number; bedrooms?: number; bathrooms?: number; square_feet?: number; lot_acres?: number; year_built?: number;
  condition?: string; sale_price?: number; sale_date?: string; monthly_rent?: number; listing_date?: string; status?: string;
  price_per_square_foot?: number; price_per_unit?: number; rent_per_unit?: number; rent_per_square_foot?: number;
  match_score?: number; score_components?: Record<string, number>; included: boolean; exclusion_reason?: string; notes?: string;
  created_at: string; updated_at: string;
};
export type ComparableInput = Omit<PropertyComparable, "comparable_id" | "subject_property_id" | "price_per_square_foot" | "price_per_unit" | "rent_per_unit" | "rent_per_square_foot" | "match_score" | "score_components" | "exclusion_reason" | "created_at" | "updated_at">;
export type ComparableAnalysis = {
  property_id: string; current_market_value_low?: number; current_market_value_high?: number; estimated_market_value?: number;
  after_repair_value_low?: number; after_repair_value_high?: number; estimated_after_repair_value?: number;
  estimated_monthly_market_rent_low?: number; estimated_monthly_market_rent_high?: number; estimated_monthly_market_rent?: number;
  estimated_annual_market_rent?: number; sale_comparable_count: number; rental_comparable_count: number;
  average_sale_distance?: number; average_rental_distance?: number; comparable_analysis_confidence: number;
  comparable_analysis_status: string; comparable_missing_items: string[]; comparable_limitations: string[];
  recommended_comparable_action: string; analyzed_at: string;
};

File: services/api/app/schemas/comparable.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, model_validator


ComparableType = Literal["sale", "rental"]


class ComparableInput(BaseModel):
    comparable_type: ComparableType
    source_name: str
    source_url: str
    source_record_id: str | None = None
    address: str
    latitude: float | None = None
    longitude: float | None = None
    distance_miles: float | None = Field(default=None, ge=0)
    property_type: str
    property_subtype: str | None = None
    unit_count: int | None = Field(default=None, ge=1)
    bedrooms: float | None = Field(default=None, ge=0)
    bathrooms: float | None = Field(default=None, ge=0)
    square_feet: float | None = Field(default=None, gt=0)
    lot_acres: float | None = Field(default=None, ge=0)
    year_built: int | None = Field(default=None, ge=1700, le=2100)
    condition: str | None = None
    sale_price: float | None = Field(default=None, gt=0)
    sale_date: date | None = None
    monthly_rent: float | None = Field(default=None, gt=0)
    listing_date: date | None = None
    status: str | None = None
    included: bool = True
    notes: str | None = None

    @model_validator(mode="after")
    def validate_evidence(self):
        if self.comparable_type == "sale" and (self.sale_price is None or self.sale_date is None):
            raise ValueError("Sale comparables require sale_price and sale_date.")
        if self.comparable_type == "rental" and (self.monthly_rent is None or self.listing_date is None):
            raise ValueError("Rental comparables require monthly_rent and listing_date.")
        return self


class ComparablePatch(BaseModel):
    source_name: str | None = None
    source_url: str | None = None
    source_record_id: str | None = None
    address: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    distance_miles: float | None = Field(default=None, ge=0)
    property_type: str | None = None
    property_subtype: str | None = None
    unit_count: int | None = Field(default=None, ge=1)
    bedrooms: float | None = Field(default=None, ge=0)
    bathrooms: float | None = Field(default=None, ge=0)
    square_feet: float | None = Field(default=None, gt=0)
    lot_acres: float | None = Field(default=None, ge=0)
    year_built: int | None = Field(default=None, ge=1700, le=2100)
    condition: str | None = None
    sale_price: float | None = Field(default=None, gt=0)
    sale_date: date | None = None
    monthly_rent: float | None = Field(default=None, gt=0)
    listing_date: date | None = None
    status: str | None = None
    included: bool | None = None
    notes: str | None = None
